packages/config/src/load.ts

```diff
@@ -0,0 +1,155 @@
+import type { ConfigResult, FlatbreadConfig } from '@flatbread/core';
+import { build } from 'esbuild';
+import colors from 'kleur';
+import fs from 'node:fs/promises';
+import path from 'node:path';
+import { pathToFileURL } from 'node:url';
+import {
+  validateConfigHasDefaultExport,
+  validateConfigStructure,
+} from './validate';
+
+type ConfigFileExtension = 'js' | 'mjs' | 'cjs' | 'ts' | 'mts' | 'cts';
+type ConfigFileName = `flatbread.config.${ConfigFileExtension}`;
+
+export const FLATBREAD_CONFIG_FILE_NAMES: ConfigFileName[] = [
+  'flatbread.config.js',
+  'flatbread.config.mjs',
+  'flatbread.config.cjs',
+
+  'flatbread.config.ts',
+  'flatbread.config.mts',
+  'flatbread.config.cts',
+];
+
+const FLATBREAD_CONFIG_FILE_REGEX = /flatbread\.config\.[mc]?[jt]s$/;
+
+const VALID_CONFIG_NAMES_MESSAGE = `Valid config filenames are:\n\t${colors.green(
+  FLATBREAD_CONFIG_FILE_NAMES.join('\n\t')
+)}
+`;
+
+/**
+ * Loads an ESModule-style config file.
+ */
+const esmLoader = async (
+  filename: string,
+  code: string
+): Promise<FlatbreadConfig> => {
+  const configModule = await loadConfigFromBundledFile(filename, code);
+
+  validateConfigStructure(configModule);
+
+  return configModule;
+};
+
+async function loadConfigFromBundledFile(
+  fileName: string,
+  bundledCode: string
+): Promise<FlatbreadConfig> {
+  // swiped a bit from https://github.com/vitejs/vite/blob/main/packages/vite/src/node/config.ts
+  //
+  // for esm, before we can register loaders without requiring users to run node
+  // with --experimental-loader themselves, we have to do a hack here:
+  // write it to disk, load it with native Node ESM, then delete the file.
+  const fileBase = `${fileName}.timestamp-${Date.now()}`;
```

Comment:
it might be worth creating a `.flatbread` folder to save these in, then we could use a hash instead of a timestamp and do caching too,  in the finally could unlink all files in that folder except filenametmp

Reply:
If we cache the bundled config file, it'll be difficult hard to know when to invalidate it which is why I suspect Vite does not cache its bundled config. Say you update a dependency that's imported in the config file, if we hash on the config file contents, nothing will change and it'll incorrectly use the old bundle.

FWIW, the unlink here is invoked almost instantly - it happens so fast that the temp file doesn't even flicker into the file tree 😛

I agree that having a `.flatbread` folder would make sense for any type of cache persistence when we get to that point, but I think we should first focus on areas that will see a higher return in caching efforts, such as deriving the types of all the fields in each collection

Reply:
oh. does this actually bundle all the imports? I was assuming it just transpiled the config file alone

Reply:
yeah it bundles all the imports